### ISSUE TITLE
instance.defaults changes global object

### ISSUE THREAD

**Issue by isterin** [axios/axios#385]
If I have two instances of axios created with axios.create and I set a default header on one, it also changes it on all the other ones in existence.

```
var instance1 = this.oauthService = axios.create({});
var instance2 = this.oauthService = axios.create({});

instance2.defaults.headers.common['Authorization'] = 'askdjfaksdjf';
```

Both instance1 and 2 now have the Authorization header changed.  I believe the problem is that the util.merge operation doesn't do deep merging, though as you see below, only if the result[key] and val are both objects, does it do a recursive merge, otherwise it assigns the value by reference.  

```
function merge(/\* obj1, obj2, obj3, ... */) {
  var result = {};
  function assignValue(val, key) {
    if (typeof result[key] === 'object' && typeof val === 'object') {
      result[key] = merge(result[key], val);
    } else {
      result[key] = val;
    }
  }

  for (var i = 0, l = arguments.length; i < l; i++) {
    forEach(arguments[i], assignValue);
  }
  return result;
}
```

**Comment by isterin** [comment]
BTW, easily fixed with lodash.cloneDeep when merging.  I'm not sure though you want to bring in lodash dependency or implement deep cloning outside of it?

**Comment by nickuraltsev** [comment]
Thank you! This definitely needs to be fixed.

**Comment by isterin** [comment]
Is bringing in outside deps like lodash acceptable?

**Comment by mzabriskie** [comment]
I played around with using `_.cloneDeep` inside the `utils.merge` and it certainly fixed this problem. However it also somehow broke stuff for node. I'm looking into what's going on there.

**Comment by nickuraltsev** [comment]
@mzabriskie Please take a look at the discussion in #370

**Comment by jonjaques** [comment]
I was able to work around this by adding a request interceptor. I have a custom class wrapping axios, so something like this:

```
    this.client.interceptors.request.use(request => {
      request.headers['Authorization'] = `Bearer ${this.token}`
      return request
    })
```

**Comment by codeclown** [comment]
Note that instead of `_.cloneDeep` or other external dependency, `JSON.parse(JSON.stringify(detaults))` may work nicely, assuming there are no circular references.

**Comment by yanivefraim** [comment]
Is it possible that this issue was already fixed? I cannot reproduce it, unless i'm missing something here. I wrote the following test, and it is green:

```js
it('should not be affected by change to another instance defaults', function() {
  var instance1 = axios.create();
  var instance2 = axios.create();

  instance1.defaults.baseURL = 'http://instance1.example.com/';
  expect(instance2.defaults.baseURL).not.toBe('http://instance1.example.com/');
});
```

I then tried the following [code](http://requirebin.com/?gist=26953e7eb0748c9ed7d5701a46bb7998), and again, defaults are not passed to other instance.

**Comment by matiishyn** [comment]
looks like the issue is still present in the newest version

**Comment by matiishyn** [comment]
so how can we fix this with `cloneDeep`?

here's what I tried, but no luck:

```js
// BASE URL
axios.defaults.baseURL = CONFIG.API_SERVER;

// SEPARATE INSTANCES
const authAxios = axios.create();
authAxios.defaults = _.cloneDeep(authAxios.defaults);

const apiAxios = axios.create();
apiAxios.defaults = _.cloneDeep(apiAxios.defaults);

// SETTING JWT HEADER
const savedToken = LS.getJwt();
if (savedToken) {
  apiAxios.defaults.headers.common.Authorization = `JWT ${savedToken}`;    
  // ^=== this will set headers to both instances :( 
}

console.log(authAxios.defaults);
console.log(apiAxios.defaults);
```

But both `authAxios` and `apiAxios` instances have the `Authorization` headers.

Did anyone fix it somehow?

**Comment by lukaszgrolik** [comment]
@matiishyn try this:
```js
if (savedToken) {
  const headers = _.cloneDeep(apiAxios.defaults.headers);
  headers.common.Authorization = `JWT ${savedToken}`;
  apiAxios.defaults.headers = headers;
}
```
Cloning and setting whole headers object works for me. Took hours of debugging and finding this thread though, so I look forward to see this fixed in the future releases.

**Comment by prakashgp** [comment]
This worked for me 
`request.defaults.headers = Object.assign({}, request.defaults.headers, {'X-Customer-Id': '123'})`

**Comment by foundryspatial-duncan** [comment]
Using `Object.assign` won't always work because it [doesn't do a "deep clone."](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Object/assign#Deep_Clone)
A cheap alternative without using lodash might be to JSON encode and decode it:

``` javascript
// save the default headers before changing any instances
const defaultHeaders = JSON.parse(JSON.stringify(axios.defaults.headers));

// change something on your instance(s)
fooInstance.defaults.headers.common.bar = "baz";

// re-set the default settings in the global axios
axios.defaults.headers = defaultHeaders;
```

This workflow should suffice until the bug is fixed, I hope!

**Comment by marbemac** [comment]
Using Axios in a server side rendering situation (next.js, meteor, any SSR focused library) leads to a huge security issue when using defaults.headers to set the currently logged in user token... this should be an urgent and critical bug IMHO. We only just discovered this, crazy its been open for over a year :/.

**Comment by MartinGian** [comment]
I use the @jonjaques solution and works fine!

**Comment by codeclown** [comment]
There is a PR open to fix this issue but it's stuck currently due to no interaction from project owners. https://github.com/axios/axios/pull/533

**Comment by robaxelsen** [comment]
@codeclown if you look at the issue you linked to, you will see that @nickuraltsev came with PR review feedback that was never acted on by @yanivefraim.

While this project currently has a lot of old issues, there was a recent popular ["help wanted for axios"](http://www.mattzabriskie.com/blog/axios-help-wanted) by the creator or Axios. Let's hope that will help clear old issues and increase response time for new ones.

Also, if you feel you can contribute, please let us know.

**Comment by yanivefraim** [comment]
@robaxelsen - as I said, I would really love to help here!

**Comment by codeclown** [comment]
@robaxelsen I understand, and I sincerely apologize for seeming hostile towards the project owners. It appeared to me that the comment on the issue did not involve direct review/instructions for the PR maintainer to act upon, but rather a wider context which can also be found on the issue https://github.com/axios/axios/issues/812.

I just looked into the problem myself in an attempt to provide a PR and discovered that the issue https://github.com/axios/axios/issues/812 really captures the tricky situation we are in due to wanting to get rid of merging all the options.

**Comment by SzymonLisowiec** [comment]
I found simple solution. While creating new instance only need to add empty `headers`.
```
let api = axios.create({
	baseURL: 'http://localhost:8080/',
	headers: {
			
	}
});

api.defaults.headers['Authorization'] = 'Bearer xxx';
```

**Comment by sony-mathew** [comment]
I think this problem is because headers for global config is named `axios.defaults.headers.common` and the same thing is accessible via `instance.defaults.headers.common`.
If you want instance level headers you don't need to use `defaults.headers.common`. You can directly use `instance.defaults.headers['X-My-Header'] = 'XXX'`.
@SzymonLisowiec has given a nice code sample in the above answer.

**Comment by heisian** [comment]
@nickuraltsev 
> Thank you! This definitely needs to be fixed.

How's that comin, eh?

**Comment by wlingke** [comment]
This is a serious bug. Can we prioritize fixing it? It's been over a year and half

**Comment by codeclown** [comment]
@heisian @wlingke This is not a simple issue to fix. Currently it is intended behaviour that Axios keeps a reference to the global defaults, as seen in [this test](https://github.com/axios/axios/blob/6d0e19343a6180fb8f5c50d843dd15feeb1f49f3/test/specs/defaults.spec.js#L151):

```js
it('should be used by custom instance if set after instance created', function (done) {
  var instance = axios.create();
  axios.defaults.baseURL = 'http://example.org/';

  instance.get('/foo');

  getAjaxRequest().then(function (request) {
    expect(request.url).toBe('http://example.org/foo');
    done();
  });
});
```

This makes it impossible to have the following test passing:
```js
it('should not be affected by changes to global defaults', function (done) {
  var instance = axios.create();
  axios.defaults.headers['X-Foobar'] = 'test';
  expect(instance.defaults.headers['X-Foobar']).toNotBe('test');
});
```

Well, it would be possible with some specific logic (e.g. if instance-specific `baseURL === null`, grab from global defaults), but I would rather see less specific logic and get rid of this restriction posed by keeping reference to global defaults.

**Comment by heisian** [comment]
@codeclown The current behaviour doesn't make any sense. Why is the reference **kept**?

One would _think_ that calling `.create` creates a new instance, and any configuration defaults applied to that instance are localized to it, not arbitrarily setting a mutable global instance. This is flawed logic.

If the `defaults` property was **truly** meant to be a global reference, then it should be immutable from the child instance. There should **also** be a separate property, like `config`, in the child instance that holds defaults for the child instance.

Currently I cannot do:
```node
myClient = axios.create()
myClient.baseURL = 'https://google.com' // Doesn't work.
```

I'm not sure this is not a "simple" issue, in fact it is a pretty basic issue for pretty basic functionality that many other languages simply leverage `curl` for, alas, everyone in the node ecosystem wants to reinvent the wheel. /rant

**Comment by codeclown** [comment]
@heisian 

> The current behaviour doesn't make any sense. Why is the reference kept?
> 
> One would think that calling .create creates a new instance, and any configuration defaults applied to that instance are localized to it, not arbitrarily setting a mutable global instance. This is flawed logic.

I agree completely. I'm not sure where to propose deprecating this "functionality", because there are so many issues regarding this defaults-issue. I do, however, think that it should be deprecated because it is not only illogical but harmful and even dangerous (especially on the server-side).

---

Anyways, whenever that problem is resolved, the second half of fixing this defaults-issue is noticing that not all the configuration can be just deep merged with the defaults. In a [comment to another issue](https://github.com/axios/axios/issues/812#issuecomment-360220938) (expand the second section) I categorized the configuration properties based on how they should be defined when a new instance is created. I hope this list is of help to anyone who might be stabbing away at this later.

**Comment by codeclown** [comment]
@heisian 
Here's how the options for an individual request are set:

https://github.com/axios/axios/blob/d4dc124f15c8dc69861b9cf92c059dbda46ae565/lib/core/Axios.js#L35

So, the problem comes from nested objects, e.g. `headers`, not primitive values such as `timeout`.

---

> at the end of the day we haven't heard from the owner in a year and a half so at best we're stuck with forks only even if there is a definite fix.

I'm sure that a PR would be very appreciated by many, considering the amount of times this issue has been reported. I personally got stuck with that one failing test. I would encourage you or anyone to create a PR if you are able to propose a fix.

**Comment by heisian** [comment]
@codeclown Got it. I'll give it a shot.

**Comment by marbemac** [comment]
Regardless of the conclusion on this issue, in the interim there should be a big bold notice on the readme that warns NOT to use defaults when in a server context (SSR w react, for example) - because it can result in bad security issues w shared API keys etc.

**Comment by heisian** [comment]
@codeclown #1391

**Comment by throrin19** [comment]
Same problem here

**Comment by nickuraltsev** [comment]
Fixed via #1395
